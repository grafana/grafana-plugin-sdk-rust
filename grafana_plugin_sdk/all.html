<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-7dda4a46.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="grafana_plugin_sdk" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (092a284ba 2025-04-13)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../static.files/storage-4e99c027.js"></script><script defer src="../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod sys"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../grafana_plugin_sdk/index.html">grafana_<wbr>plugin_<wbr>sdk</a><span class="version">0.5.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h3><a href="#structs">Crate Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#constants" title="Constants">Constants</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="backend/struct.AppInstanceSettings.html">backend::AppInstanceSettings</a></li><li><a href="backend/struct.AppPlugin.html">backend::AppPlugin</a></li><li><a href="backend/struct.CheckHealthResponse.html">backend::CheckHealthResponse</a></li><li><a href="backend/struct.CollectMetricsResponse.html">backend::CollectMetricsResponse</a></li><li><a href="backend/struct.ConfigError.html">backend::ConfigError</a></li><li><a href="backend/struct.DataQuery.html">backend::DataQuery</a></li><li><a href="backend/struct.DataResponse.html">backend::DataResponse</a></li><li><a href="backend/struct.DataSourceInstanceSettings.html">backend::DataSourceInstanceSettings</a></li><li><a href="backend/struct.DataSourcePlugin.html">backend::DataSourcePlugin</a></li><li><a href="backend/struct.GrafanaConfig.html">backend::GrafanaConfig</a></li><li><a href="backend/struct.HCLogJson.html">backend::HCLogJson</a></li><li><a href="backend/struct.InitialData.html">backend::InitialData</a></li><li><a href="backend/struct.MetricsPayload.html">backend::MetricsPayload</a></li><li><a href="backend/struct.Plugin.html">backend::Plugin</a></li><li><a href="backend/struct.PluginContext.html">backend::PluginContext</a></li><li><a href="backend/struct.PublishStreamResponse.html">backend::PublishStreamResponse</a></li><li><a href="backend/struct.StreamPacket.html">backend::StreamPacket</a></li><li><a href="backend/struct.SubscribeStreamResponse.html">backend::SubscribeStreamResponse</a></li><li><a href="backend/struct.TimeRange.html">backend::TimeRange</a></li><li><a href="backend/struct.User.html">backend::User</a></li><li><a href="data/struct.CheckedFrame.html">data::CheckedFrame</a></li><li><a href="data/struct.ConfFloat64.html">data::ConfFloat64</a></li><li><a href="data/struct.DataLink.html">data::DataLink</a></li><li><a href="data/struct.Field.html">data::Field</a></li><li><a href="data/struct.FieldConfig.html">data::FieldConfig</a></li><li><a href="data/struct.Frame.html">data::Frame</a></li><li><a href="data/struct.Metadata.html">data::Metadata</a></li><li><a href="data/struct.Notice.html">data::Notice</a></li><li><a href="data/struct.QueryStat.html">data::QueryStat</a></li><li><a href="data/struct.Threshold.html">data::Threshold</a></li><li><a href="data/struct.ThresholdsConfig.html">data::ThresholdsConfig</a></li><li><a href="data/struct.TypeInfo.html">data::TypeInfo</a></li><li><a href="data/struct.ValueMappingResult.html">data::ValueMappingResult</a></li><li><a href="live/struct.Channel.html">live::Channel</a></li><li><a href="live/struct.Namespace.html">live::Namespace</a></li><li><a href="live/struct.Path.html">live::Path</a></li><li><a href="pluginv2/struct.AdmissionRequest.html">pluginv2::AdmissionRequest</a></li><li><a href="pluginv2/struct.AppInstanceSettings.html">pluginv2::AppInstanceSettings</a></li><li><a href="pluginv2/struct.CallResourceRequest.html">pluginv2::CallResourceRequest</a></li><li><a href="pluginv2/struct.CallResourceResponse.html">pluginv2::CallResourceResponse</a></li><li><a href="pluginv2/struct.CheckHealthRequest.html">pluginv2::CheckHealthRequest</a></li><li><a href="pluginv2/struct.CheckHealthResponse.html">pluginv2::CheckHealthResponse</a></li><li><a href="pluginv2/struct.CollectMetricsRequest.html">pluginv2::CollectMetricsRequest</a></li><li><a href="pluginv2/struct.CollectMetricsResponse.html">pluginv2::CollectMetricsResponse</a></li><li><a href="pluginv2/struct.ConversionRequest.html">pluginv2::ConversionRequest</a></li><li><a href="pluginv2/struct.ConversionResponse.html">pluginv2::ConversionResponse</a></li><li><a href="pluginv2/struct.DataQuery.html">pluginv2::DataQuery</a></li><li><a href="pluginv2/struct.DataResponse.html">pluginv2::DataResponse</a></li><li><a href="pluginv2/struct.DataSourceInstanceSettings.html">pluginv2::DataSourceInstanceSettings</a></li><li><a href="pluginv2/struct.GroupVersion.html">pluginv2::GroupVersion</a></li><li><a href="pluginv2/struct.GroupVersionKind.html">pluginv2::GroupVersionKind</a></li><li><a href="pluginv2/struct.MutationResponse.html">pluginv2::MutationResponse</a></li><li><a href="pluginv2/struct.PluginContext.html">pluginv2::PluginContext</a></li><li><a href="pluginv2/struct.PublishStreamRequest.html">pluginv2::PublishStreamRequest</a></li><li><a href="pluginv2/struct.PublishStreamResponse.html">pluginv2::PublishStreamResponse</a></li><li><a href="pluginv2/struct.QueryDataRequest.html">pluginv2::QueryDataRequest</a></li><li><a href="pluginv2/struct.QueryDataResponse.html">pluginv2::QueryDataResponse</a></li><li><a href="pluginv2/struct.RawObject.html">pluginv2::RawObject</a></li><li><a href="pluginv2/struct.RunStreamRequest.html">pluginv2::RunStreamRequest</a></li><li><a href="pluginv2/struct.StatusResult.html">pluginv2::StatusResult</a></li><li><a href="pluginv2/struct.StreamPacket.html">pluginv2::StreamPacket</a></li><li><a href="pluginv2/struct.StringList.html">pluginv2::StringList</a></li><li><a href="pluginv2/struct.SubscribeStreamRequest.html">pluginv2::SubscribeStreamRequest</a></li><li><a href="pluginv2/struct.SubscribeStreamResponse.html">pluginv2::SubscribeStreamResponse</a></li><li><a href="pluginv2/struct.TimeRange.html">pluginv2::TimeRange</a></li><li><a href="pluginv2/struct.User.html">pluginv2::User</a></li><li><a href="pluginv2/struct.ValidationResponse.html">pluginv2::ValidationResponse</a></li><li><a href="pluginv2/admission_control_client/struct.AdmissionControlClient.html">pluginv2::admission_control_client::AdmissionControlClient</a></li><li><a href="pluginv2/admission_control_server/struct.AdmissionControlServer.html">pluginv2::admission_control_server::AdmissionControlServer</a></li><li><a href="pluginv2/collect_metrics_response/struct.Payload.html">pluginv2::collect_metrics_response::Payload</a></li><li><a href="pluginv2/data_client/struct.DataClient.html">pluginv2::data_client::DataClient</a></li><li><a href="pluginv2/data_server/struct.DataServer.html">pluginv2::data_server::DataServer</a></li><li><a href="pluginv2/diagnostics_client/struct.DiagnosticsClient.html">pluginv2::diagnostics_client::DiagnosticsClient</a></li><li><a href="pluginv2/diagnostics_server/struct.DiagnosticsServer.html">pluginv2::diagnostics_server::DiagnosticsServer</a></li><li><a href="pluginv2/resource_client/struct.ResourceClient.html">pluginv2::resource_client::ResourceClient</a></li><li><a href="pluginv2/resource_conversion_client/struct.ResourceConversionClient.html">pluginv2::resource_conversion_client::ResourceConversionClient</a></li><li><a href="pluginv2/resource_conversion_server/struct.ResourceConversionServer.html">pluginv2::resource_conversion_server::ResourceConversionServer</a></li><li><a href="pluginv2/resource_server/struct.ResourceServer.html">pluginv2::resource_server::ResourceServer</a></li><li><a href="pluginv2/stream_client/struct.StreamClient.html">pluginv2::stream_client::StreamClient</a></li><li><a href="pluginv2/stream_server/struct.StreamServer.html">pluginv2::stream_server::StreamServer</a></li></ul><h3 id="enums">Enums</h3><ul class="all-items"><li><a href="backend/enum.ConvertFromError.html">backend::ConvertFromError</a></li><li><a href="backend/enum.ConvertToError.html">backend::ConvertToError</a></li><li><a href="backend/enum.DataQueryStatus.html">backend::DataQueryStatus</a></li><li><a href="backend/enum.Error.html">backend::Error</a></li><li><a href="backend/enum.ErrorSource.html">backend::ErrorSource</a></li><li><a href="backend/enum.HealthStatus.html">backend::HealthStatus</a></li><li><a href="backend/enum.PublishStreamStatus.html">backend::PublishStreamStatus</a></li><li><a href="backend/enum.Role.html">backend::Role</a></li><li><a href="backend/enum.SubscribeStreamStatus.html">backend::SubscribeStreamStatus</a></li><li><a href="data/enum.Error.html">data::Error</a></li><li><a href="data/enum.FrameInclude.html">data::FrameInclude</a></li><li><a href="data/enum.InspectType.html">data::InspectType</a></li><li><a href="data/enum.Severity.html">data::Severity</a></li><li><a href="data/enum.SimpleType.html">data::SimpleType</a></li><li><a href="data/enum.SpecialValueMatch.html">data::SpecialValueMatch</a></li><li><a href="data/enum.ThresholdsMode.html">data::ThresholdsMode</a></li><li><a href="data/enum.TypeInfoType.html">data::TypeInfoType</a></li><li><a href="data/enum.ValueMapping.html">data::ValueMapping</a></li><li><a href="data/enum.VisType.html">data::VisType</a></li><li><a href="live/enum.ChannelError.html">live::ChannelError</a></li><li><a href="live/enum.Scope.html">live::Scope</a></li><li><a href="pluginv2/admission_request/enum.Operation.html">pluginv2::admission_request::Operation</a></li><li><a href="pluginv2/check_health_response/enum.HealthStatus.html">pluginv2::check_health_response::HealthStatus</a></li><li><a href="pluginv2/publish_stream_response/enum.Status.html">pluginv2::publish_stream_response::Status</a></li><li><a href="pluginv2/subscribe_stream_response/enum.Status.html">pluginv2::subscribe_stream_response::Status</a></li></ul><h3 id="traits">Traits</h3><ul class="all-items"><li><a href="backend/trait.DataQueryError.html">backend::DataQueryError</a></li><li><a href="backend/trait.DataService.html">backend::DataService</a></li><li><a href="backend/trait.DiagnosticsService.html">backend::DiagnosticsService</a></li><li><a href="backend/trait.ErrIntoHttpResponse.html">backend::ErrIntoHttpResponse</a></li><li><a href="backend/trait.GrafanaPlugin.html">backend::GrafanaPlugin</a></li><li><a href="backend/trait.InstanceSettings.html">backend::InstanceSettings</a></li><li><a href="backend/trait.IntoHttpResponse.html">backend::IntoHttpResponse</a></li><li><a href="backend/trait.PluginType.html">backend::PluginType</a></li><li><a href="backend/trait.ResourceService.html">backend::ResourceService</a></li><li><a href="backend/trait.StreamService.html">backend::StreamService</a></li><li><a href="data/trait.ArrayIntoField.html">data::ArrayIntoField</a></li><li><a href="data/trait.ArrayRefIntoField.html">data::ArrayRefIntoField</a></li><li><a href="data/trait.FieldType.html">data::FieldType</a></li><li><a href="data/trait.FromFields.html">data::FromFields</a></li><li><a href="data/trait.IntoField.html">data::IntoField</a></li><li><a href="data/trait.IntoFieldType.html">data::IntoFieldType</a></li><li><a href="data/trait.IntoFrame.html">data::IntoFrame</a></li><li><a href="data/trait.IntoOptField.html">data::IntoOptField</a></li><li><a href="pluginv2/admission_control_server/trait.AdmissionControl.html">pluginv2::admission_control_server::AdmissionControl</a></li><li><a href="pluginv2/data_server/trait.Data.html">pluginv2::data_server::Data</a></li><li><a href="pluginv2/diagnostics_server/trait.Diagnostics.html">pluginv2::diagnostics_server::Diagnostics</a></li><li><a href="pluginv2/resource_conversion_server/trait.ResourceConversion.html">pluginv2::resource_conversion_server::ResourceConversion</a></li><li><a href="pluginv2/resource_server/trait.Resource.html">pluginv2::resource_server::Resource</a></li><li><a href="pluginv2/stream_server/trait.Stream.html">pluginv2::stream_server::Stream</a></li></ul><h3 id="attributes">Attribute Macros</h3><ul class="all-items"><li><a href="backend/attr.async_trait.html">backend::async_trait</a></li><li><a href="attr.main.html">main</a></li></ul><h3 id="functions">Functions</h3><ul class="all-items"><li><a href="backend/fn.initialize.html">backend::initialize</a></li><li><a href="backend/fn.layer.html">backend::layer</a></li></ul><h3 id="types">Type Aliases</h3><ul class="all-items"><li><a href="backend/type.BoxDataResponseStream.html">backend::BoxDataResponseStream</a></li><li><a href="backend/type.BoxResourceFuture.html">backend::BoxResourceFuture</a></li><li><a href="backend/type.BoxResourceStream.html">backend::BoxResourceStream</a></li><li><a href="backend/type.BoxRunStream.html">backend::BoxRunStream</a></li><li><a href="backend/type.CallResourceRequest.html">backend::CallResourceRequest</a></li><li><a href="backend/type.CheckHealthRequest.html">backend::CheckHealthRequest</a></li><li><a href="backend/type.CollectMetricsRequest.html">backend::CollectMetricsRequest</a></li><li><a href="backend/type.PublishStreamRequest.html">backend::PublishStreamRequest</a></li><li><a href="backend/type.QueryDataRequest.html">backend::QueryDataRequest</a></li><li><a href="backend/type.RunStreamRequest.html">backend::RunStreamRequest</a></li><li><a href="backend/type.SubscribeStreamRequest.html">backend::SubscribeStreamRequest</a></li></ul><h3 id="constants">Constants</h3><ul class="all-items"><li><a href="data/constant.EXPLICIT_NULL.html">data::EXPLICIT_NULL</a></li><li><a href="data/constant.TIME_FIELD_NAME.html">data::TIME_FIELD_NAME</a></li><li><a href="data/constant.VALUE_FIELD_NAME.html">data::VALUE_FIELD_NAME</a></li><li><a href="live/constant.MAX_CHANNEL_LENGTH.html">live::MAX_CHANNEL_LENGTH</a></li><li><a href="pluginv2/admission_control_server/constant.SERVICE_NAME.html">pluginv2::admission_control_server::SERVICE_NAME</a></li><li><a href="pluginv2/data_server/constant.SERVICE_NAME.html">pluginv2::data_server::SERVICE_NAME</a></li><li><a href="pluginv2/diagnostics_server/constant.SERVICE_NAME.html">pluginv2::diagnostics_server::SERVICE_NAME</a></li><li><a href="pluginv2/resource_conversion_server/constant.SERVICE_NAME.html">pluginv2::resource_conversion_server::SERVICE_NAME</a></li><li><a href="pluginv2/resource_server/constant.SERVICE_NAME.html">pluginv2::resource_server::SERVICE_NAME</a></li><li><a href="pluginv2/stream_server/constant.SERVICE_NAME.html">pluginv2::stream_server::SERVICE_NAME</a></li></ul></section></div></main></body></html>