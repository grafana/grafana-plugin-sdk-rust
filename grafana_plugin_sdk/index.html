<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The Grafana Plugin SDK for Rust."><title>grafana_plugin_sdk - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-7dda4a46.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="grafana_plugin_sdk" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (092a284ba 2025-04-13)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../static.files/storage-4e99c027.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../grafana_plugin_sdk/index.html">grafana_<wbr>plugin_<wbr>sdk</a><span class="version">0.5.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#feature-flags" title="Feature flags">Feature flags</a></li></ul><h3><a href="#reexports">Crate Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>grafana_plugin_sdk</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/grafana_plugin_sdk/lib.rs.html#1-85">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The Grafana Plugin SDK for Rust.</p>
<p>This crate contains a Rust implementation of the Grafana plugin SDK. It is divided into three main modules:</p>
<ul>
<li><a href="backend/index.html" title="mod grafana_plugin_sdk::backend"><code>backend</code></a> contains the traits that must be implemented by backend plugins for various pieces of functionality,
whether querying data, calling resources, or streaming data between Grafana and the plugin.</li>
<li><a href="data/index.html" title="mod grafana_plugin_sdk::data"><code>data</code></a> contains the fundamental data structures used by backend plugins, such as <a href="data/struct.Frame.html" title="struct grafana_plugin_sdk::data::Frame"><code>Frame</code></a>s, <a href="data/struct.Field.html" title="struct grafana_plugin_sdk::data::Field"><code>Field</code></a>s,
and their associated metadata.</li>
<li><a href="live/index.html" title="mod grafana_plugin_sdk::live"><code>live</code></a> contains functionality used by <a href="https://grafana.com/docs/grafana/latest/live/">Grafana Live</a>, the streaming messaging service available from
Grafana 8.0.</li>
</ul>
<p>The <a href="prelude/index.html" title="mod grafana_plugin_sdk::prelude"><code>prelude</code></a> contains some useful unambiguous traits which are helpful when creating some structures,
particularly <a href="data/struct.Frame.html" title="struct grafana_plugin_sdk::data::Frame"><code>Frame</code></a>s and <a href="data/struct.Field.html" title="struct grafana_plugin_sdk::data::Field"><code>Field</code></a>s.</p>
<p>Backend plugins communicate with Grafana via gRPC. The low-level protocols are exposed in the <a href="pluginv2/index.html" title="mod grafana_plugin_sdk::pluginv2"><code>pluginv2</code></a>
module as an escape hatch, if required. Please file an issue if this is needed and we will try to
accommodate your needs in the next release of the high-level SDK.</p>
<p>See the docs on <a href="https://grafana.com/docs/grafana/latest/developers/plugins/backend/">backend plugins on grafana.com</a> for an introduction to backend Grafana plugins, or check out
the <a href="https://github.com/grafana/grafana-plugin-sdk-rust/tree/main/examples">crate examples</a> or <a href="https://github.com/sd2k/grafana-sample-backend-plugin-rust/">sample app repo</a> to get started with writing a backend plugin in Rust.</p>
<h2 id="feature-flags"><a class="doc-anchor" href="#feature-flags">ยง</a>Feature flags</h2>
<p>The following feature flags enable additional functionality for this crate:</p>
<ul>
<li><code>reqwest</code> - adds an <a href="backend/trait.IntoHttpResponse.html" title="trait grafana_plugin_sdk::backend::IntoHttpResponse"><code>IntoHttpResponse</code></a> implementation for
[<code>reqwest::Response</code>]</li>
</ul>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">ยง</a></h2><dl class="item-table reexports"><dt id="reexport.arrow"><code>pub use arrow;</code></dt></dl><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="backend/index.html" title="mod grafana_plugin_sdk::backend">backend</a></dt><dd>Functionality for use by backend plugins.</dd><dt><a class="mod" href="data/index.html" title="mod grafana_plugin_sdk::data">data</a></dt><dd>Data types used throughout the SDK.</dd><dt><a class="mod" href="live/index.html" title="mod grafana_plugin_sdk::live">live</a></dt><dd>Grafana Live features, providing streaming functionality to Grafana.</dd><dt><a class="mod" href="pluginv2/index.html" title="mod grafana_plugin_sdk::pluginv2">pluginv2</a></dt><dd>The low-level structs generated from protocol definitions.</dd><dt><a class="mod" href="prelude/index.html" title="mod grafana_plugin_sdk::prelude">prelude</a></dt><dd>Contains useful helper traits for constructing <a href="data/struct.Field.html" title="struct grafana_plugin_sdk::data::Field"><code>Field</code></a>s and <a href="data/struct.Frame.html" title="struct grafana_plugin_sdk::data::Frame"><code>Frame</code></a>s.</dd></dl><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="attr" href="attr.main.html" title="attr grafana_plugin_sdk::main">main</a></dt><dd>Generates a <code>main</code> function that starts a <a href="../grafana_plugin_sdk/backend/struct.Plugin.html"><code>Plugin</code></a> with the returned service struct.</dd></dl></section></div></main></body></html>