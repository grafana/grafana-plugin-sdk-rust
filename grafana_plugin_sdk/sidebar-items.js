window.SIDEBAR_ITEMS = {"attr":["main"],"mod":["backend","data","live","pluginv2","prelude"]};