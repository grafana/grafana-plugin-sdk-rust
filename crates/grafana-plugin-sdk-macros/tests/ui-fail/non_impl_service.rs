#![allow(dead_code, unused_variables)]

mod a {
    use grafana_plugin_sdk::prelude::*;

    #[derive(Clone, GrafanaPlugin)]
    #[grafana_plugin(plugin_type = "datasource")]
    struct MyPlugin;

    #[grafana_plugin_sdk::main(services(data))]
    async fn plugin() -> MyPlugin {
        MyPlugin
    }
}

fn main() {}
