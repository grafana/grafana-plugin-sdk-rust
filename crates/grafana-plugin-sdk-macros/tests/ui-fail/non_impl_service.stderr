error[E0277]: the trait bound `MyPlugin: DataService` is not satisfied
 --> tests/ui-fail/non_impl_service.rs:7:5
  |
7 |     #[grafana_plugin_sdk::main(services(data))]
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `DataService` is not implemented for `MyPlugin`
  |
  = help: the trait `DataService` is implemented for `backend::noop::NoopService`
note: required by a bound in `grafana_plugin_sdk::backend::Plugin::<D, Q, R, S>::data_service`
 --> $WORKSPACE/crates/grafana-plugin-sdk/src/backend/mod.rs
  |
  |     pub fn data_service<T>(self, service: T) -> Plugin<D, T, R, S>
  |            ------------ required by a bound in this associated function
  |     where
  |         T: DataService + Send + Sync + 'static,
  |            ^^^^^^^^^^^ required by this bound in `Plugin::<D, Q, R, S>::data_service`
  = note: this error originates in the attribute macro `grafana_plugin_sdk::main` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `start` exists for struct `Plugin<NoopService, MyPlugin, NoopService, NoopService>`, but its trait bounds were not satisfied
 --> tests/ui-fail/non_impl_service.rs:7:5
  |
5 |     struct MyPlugin;
  |     --------------- doesn't satisfy `MyPlugin: DataService`
6 |
7 |     #[grafana_plugin_sdk::main(services(data))]
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method cannot be called due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `MyPlugin: DataService`
note: the trait `DataService` must be implemented
 --> $WORKSPACE/crates/grafana-plugin-sdk/src/backend/data.rs
  |
  | pub trait DataService {
  | ^^^^^^^^^^^^^^^^^^^^^
  = note: this error originates in the attribute macro `grafana_plugin_sdk::main` (in Nightly builds, run with -Z macro-backtrace for more info)
